feat: support count min sketch data structure and most commands

jonathanc-n opened the pull request:
references issue: #2425 

The commands, IncrBy, Info, InitByDim, InitByProb, and Query have been made.

The merge command will still need working on, and go integration tests will be added for the future. I will also probably add a cache similar to the hyperloglog for in-memory operations.

jonathanc-n committed:
CountMinSketch Additions

jonathanc-n committed:
adding redis_cms

mapleFU commented on src/commands/commander.h:
```diff
@@ -73,6 +73,7 @@ enum class CommandCategory : uint8_t {
   Bit,
   BloomFilter,
   Cluster,
+  CMS,
```
Would this better being CMSketh? Or `CountMinSketch`?

jonathanc-n replied:
I think CMS is fine, CountMinSketch would be too long. HyperLogLog is also written as HLL here, so it is consistent.

mapleFU commented on src/types/cms.h:
```diff
@@ -0,0 +1,82 @@
+#include <assert.h>
+
+#include <atomic>
+#include <iostream>
```
Why?

mapleFU commented on src/types/cms.h:
```diff
@@ -0,0 +1,82 @@
+
+class CMSketch {
+ public:
+  explicit CMSketch(uint32_t width = 0, uint32_t depth = 0, uint64_t counter = 0, std::vector<uint32_t> array = {})
```
Why using default args?

mapleFU commented on src/types/cms.h:
```diff
@@ -0,0 +1,82 @@
+        counter_(counter),
+        array_(array.empty() ? std::vector<uint32_t>(width * depth, 0) : std::move(array)) {}
+
+  ~CMSketch() = default;
```
Remove this?

mapleFU commented on src/types/cms.h:
```diff
@@ -0,0 +1,82 @@
+
+  ~CMSketch() = default;
+
+  static CMSketch* NewCMSketch(size_t width, size_t depth) { return new CMSketch(width, depth); }
```
Why call new explicitly? Shouldn't a unique_ptr being returned in this case?

mapleFU commented on src/types/cms.h:
```diff
@@ -0,0 +1,82 @@
+
+  static CMSketch* NewCMSketch(size_t width, size_t depth) { return new CMSketch(width, depth); }
+
+  static void CMSDimFromProb(double error, double delta, uint32_t& width, uint32_t& depth);
```
Would you mind encapsulate `(width, depth)` to a struct, and just return the struct?

mapleFU commented on src/types/cms.h:
```diff
@@ -0,0 +1,82 @@
+  static CMSketch* NewCMSketch(size_t width, size_t depth) { return new CMSketch(width, depth); }
+
+  static void CMSDimFromProb(double error, double delta, uint32_t& width, uint32_t& depth);
+
```
`const char* item, size_t item_len` -> can this be change to std::string_view

mapleFU commented on src/types/cms.h:
```diff
@@ -0,0 +1,82 @@
+  const uint64_t& GetCounter() const { return counter_; }
+  const std::vector<uint32_t>& GetArray() const { return array_; }
+
+  size_t GetWidth() const { return width_; }
+  size_t GetDepth() const { return depth_; }
```
Why do they shift between u32 and size_t?

mapleFU commented on src/types/cms.h:
```diff
@@ -0,0 +1,82 @@
+
+  int CMSMergeParams(const MergeParams& params);
+
+  uint64_t& GetCounter() { return counter_; }
+  std::vector<uint32_t>& GetArray() { return array_; }
```
Would a set array being better here?

jonathanc-n replied:
No, for a cms, it is storing counters and not unique objects.

mapleFU reviewed (commented):
I didn't go through the detail, just some style nits firstly

mapleFU commented on src/types/cms.cc:
```diff
@@ -0,0 +1,128 @@
+  size_t min_count = std::numeric_limits<size_t>::max();
+
+  for (size_t i = 0; i < depth_; ++i) {
+    uint32_t hash = HllMurMurHash64A(item, static_cast<int>(item_len), i);
```
I don't know getting u32 from u64 hash a good idea...

jonathanc-n replied:
@mapleFU  tyty for the review, i'll be revising it tmr.

jonathanc-n committed:
Merge branch 'unstable' into cms

jonathanc-n committed:
small changes

jonathanc-n committed:
Merge branch 'cms' of https://github.com/jonathanc-n/kvrocks into cms

jonathanc-n commented:
Was there a change to GetOptions recently? After the merge, it seems to be behaving different

jonathanc-n commented:
Was there a change made to GetOptions recently? There seems to be a problem after the merge.

mapleFU commented:
> Was there a change made to GetOptions recently? There seems to be a problem after the merge.

Yes, the operations to db requires an options here.

jonathanc-n committed:
Resolved changes

jonathanc-n commented:
I believe this pr should be good for further review

mapleFU commented:
Nice, I'll take a round tonight

mapleFU commented on src/commands/cmd_cms.cc:
```diff
@@ -0,0 +1,168 @@
+ public:
+  Status Execute(Server *srv, Connection *conn, std::string *output) override {
+    if ((args_.size() - 2) % 2 != 0) {
+      return Status::RedisTryAgain;
```
?

jonathanc-n replied:
should I make this part with a status parse instead of putting it in execute?

mapleFU replied:
When arg mismatch, try again never solving this, it should report input is invalid

Besides, the arg-parsing framework can well handle this?

jonathanc-n replied:
@mapleFU alright should be good!

mapleFU commented on src/commands/cmd_cms.cc:
```diff
@@ -0,0 +1,168 @@
+      std::string key = args_[i];
+      uint64_t value = 0;
+      try {
+        value = std::stoull(args_[i + 1]);
```
we have `ParseInt` in `parse_util.h`?

mapleFU commented on src/commands/cmd_cms.cc:
```diff
@@ -0,0 +1,168 @@
+    redis::CMS cms(srv->storage, conn->GetNamespace());
+    engine::Context ctx(srv->storage);
+    rocksdb::Status s;
+    std::unordered_map<std::string, uint64_t> elements;
```
never used?

mapleFU commented on src/commands/cmd_cms.cc:
```diff
@@ -0,0 +1,168 @@
+      return {Status::RedisExecErr, s.ToString()};
+    }
+
+    *output = redis::Array({redis::BulkString("width"), redis::Integer(ret[0]), redis::BulkString("depth"),
+                            redis::Integer(ret[1]), redis::BulkString("count"), redis::Integer(ret[2])});
```
Can you use a struct for `ret` rather than a `vector`?

mapleFU commented on src/commands/cmd_cms.cc:
```diff
@@ -0,0 +1,168 @@
+    redis::CMS cms(srv->storage, conn->GetNamespace());
+    engine::Context ctx(srv->storage);
+    rocksdb::Status s;
+    uint64_t width = std::stoull(args_[2]);
+    uint64_t depth = std::stoull(args_[3]);
```
ParseInt

mapleFU commented on src/commands/cmd_cms.cc:
```diff
@@ -0,0 +1,168 @@
+    redis::CMS cms(srv->storage, conn->GetNamespace());
+    engine::Context ctx(srv->storage);
+    rocksdb::Status s;
+    double error = std::stod(args_[2]);
+    double delta = std::stod(args_[3]);
```
ParseFloat

mapleFU commented on src/storage/redis_metadata.h:
```diff
@@ -335,3 +336,15 @@ class HyperLogLogMetadata : public Metadata {
+
+class CountMinSketchMetadata : public Metadata {
+ public:
+  uint32_t width;
+  uint32_t depth;
+  uint64_t counter;
```
Should this have default value?

mapleFU commented on src/types/cms.cc:
```diff
@@ -0,0 +1,128 @@
+  for (size_t i = 0; i < depth_; ++i) {
+    uint64_t hash = HllMurMurHash64A(item.data(), static_cast<int>(item.size()), i);
+    size_t loc = (hash % width_) + (i * width_);
+    array_[loc] += value;
+    if (array_[loc] < value) {
+      array_[loc] = UINT32_MAX;
+    }
```
can this changed to check overflow?

mapleFU commented on src/types/cms.cc:
```diff
@@ -0,0 +1,128 @@
+  return dims;
+}
+
+size_t CMSketch::IncrBy(std::string_view item, size_t value) {
```
why value is a size_t here?

mapleFU commented on src/types/cms.cc:
```diff
@@ -0,0 +1,128 @@
+CMSketch::CMSketchDimensions CMSketch::CMSDimFromProb(double error, double delta) {
+  CMSketchDimensions dims;
+  dims.width = std::ceil(2 / error);
+  dims.depth = std::ceil(std::log10(delta) / std::log10(0.5));
```
Should we also try to adding a verify to check the width and depth not too large or too small ?

mapleFU replied:
Why did this being solved? Would you mind point out the code resolve this problem?

mapleFU replied:
Let me change the wording here, can we assure:
1. the cmsketch memory used less than 50MB
2.the cmsketch memory used greater than 0

mapleFU commented on src/types/redis_cms.cc:
```diff
@@ -0,0 +1,167 @@
+  std::string ns_key = AppendNamespacePrefix(user_key);
+
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+  CountMinSketchMetadata metadata{};
```
If not exists, does this mean getting a `<0, 0, 0>` pair? What does it means?

jonathanc-n replied:
Yeah, you would get a 0,0,0 pair

mapleFU commented on src/types/cms.cc:
```diff
@@ -0,0 +1,128 @@
+
+  for (size_t i = 0; i < depth_; ++i) {
+    uint64_t hash = HllMurMurHash64A(item.data(), static_cast<int>(item.size()), i);
+    size_t loc = (hash % width_) + (i * width_);
```
can you extract the loc to a helper function?

mapleFU reviewed (commented):
I would take a look at redis sketch later

jonathanc-n committed:
Merge branch 'unstable' into cms

jonathanc-n committed:
small tweaks

jonathanc-n commented:
@mapleFU do the changes seem right? sorry for the bother.

jonathanc-n committed:
parse change

jonathanc-n committed:
Merge branch 'unstable' into cms

jonathanc-n committed:
format changes

jonathanc-n commented:
@PragmaTwice Is the workflow able to run again?

mapleFU commented:
> do the changes seem right? sorry for the bother.

Sorry for delaying, I'll take a pass

mapleFU commented on src/storage/redis_metadata.h:
```diff
@@ -335,3 +336,15 @@ class HyperLogLogMetadata : public Metadata {
+  uint32_t width;
+  uint32_t depth;
+  uint64_t counter = 0;
+  std::vector<uint32_t> array;
```
@git-hulk @PragmaTwice 

I'm thinking about handling the `array`. Since regarding the cmsketch as a string is ok to me, but parsing it to an `std::vector` is a bit weird to me. Can we take a `Buffer` and do zero copying when doing this?

Like:
```
LoadMetadata from string. When storing, forcing LittleEndian in the array
When reading from metadata, report error if not length enough, and hold a sliced buffer on underlying data
When read/write, using memcpy to avoid unaligned access.
```

PragmaTwice replied:
Yeah it can be like how we handle the JSON data structure.

PragmaTwice replied:
But here I'm more interested in whether it's better to be a single key or putting the array in subkeys.

mapleFU replied:
> But here I'm more interested in whether it's better to be a single key or putting the array in subkeys.

Actually I think this in single key is ok since we merely "only query metadata" if not calling "info"

PragmaTwice replied:
What's the limit of the size of such an array? Do we need to consider to split it?

mapleFU replied:
https://github.com/apache/kvrocks/pull/2524/files#r1779978212

@PragmaTwice The problem is that, the stored value would always be fetched together when required. It's merely to "partition querying" / "updating" since it's a prob structure

We can limit it to 50M or less?

PragmaTwice replied:
I don't know if it's good to have one rocksdb key with a 50M-bytes value.

I checked the code and seems we'll access `d` elements (once for every row of `w` elements) in the `d*w` array.
Then I think maybe it's good to keep it in one key if the total size is small enough, e.g. 1M.

And if it's larger, we can move them into `w/k` key-values (one column as one key), and calculate how many keys we need to retrieve.

mapleFU replied:
> I don't know if it's good to have one rocksdb key with a 50M-bytes value.

The default engine doesn't handle this well. There're some blog engine which could do this, and we can enable the blob in kvrocks here.

Perhaps we can limit the size to 1MB firstly? 🤔

PragmaTwice replied:
Looks good to me.

Also yeah we should not put the array inside the metadata.

mapleFU replied:
I may try a round tonight

jonathanc-n replied:
I'll adjust it to 1MB limit

jonathanc-n replied:
So do we want to hold the array in a buffer on storage, and do operations on that?

mapleFU commented:
@jonathanc-n would you mind fix the lint?

mapleFU commented on src/storage/redis_metadata.cc:
```diff
@@ -329,7 +329,8 @@ bool Metadata::ExpireAt(uint64_t expired_ts) const {
 
 bool Metadata::IsEmptyableType() const {
-  return IsSingleKVType() || Type() == kRedisStream || Type() == kRedisBloomFilter || Type() == kRedisHyperLogLog;
+  return IsSingleKVType() || Type() == kRedisStream || Type() == kRedisBloomFilter || Type() == kRedisHyperLogLog ||
```
Wait, isn't a CMSketch a Single-KV Type?

jonathanc-n replied:
Oh, sorry about that, i was thinking due to the different variables, that it would be instead put under emptyable type. But after review it makes more sense as a singlekvtype

mapleFU commented on src/types/cms.cc:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  return 0;
+}
```
fmt: `\n` after this

mapleFU commented on src/types/cms.cc:
```diff
@@ -0,0 +1,128 @@
+  size_t min_count = std::numeric_limits<size_t>::max();
+
+  for (size_t i = 0; i < depth_; ++i) {
+    uint64_t hash = HllMurMurHash64A(item.data(), static_cast<int>(item.size()), i);
```
Please, a uint64_t as hash might overflows uint32_t, making `array_` likely to randomly set uint32 max

mapleFU replied:
What about using xxhash32?

mapleFU commented on src/types/cms.cc:
```diff
@@ -0,0 +1,128 @@
+
+  for (size_t i = 0; i < depth_; ++i) {
+    uint64_t hash = HllMurMurHash64A(item.data(), static_cast<int>(item.size()), i);
+    min_count = std::min(min_count, static_cast<size_t>(array_[(hash % width_) + (i * width_)]));
```
Why didn't `GetLocation` used here?

mapleFU commented on src/types/cms.h:
```diff
@@ -0,0 +1,89 @@
+#include <vector>
+
+#include "server/redis_reply.h"
+#include "vendor/murmurhash2.h"
```
why murmurhash2 included here?

mapleFU commented on src/types/cms.h:
```diff
@@ -0,0 +1,89 @@
+        counter_(counter),
+        array_(array.empty() ? std::vector<uint32_t>(width * depth, 0) : std::move(array)) {}
+
+  static std::unique_ptr<CMSketch> NewCMSketch(uint32_t width, int32_t depth) {
```
Ever used?

mapleFU commented:
I'm outing these days, so maybe late reply

jonathanc-n committed:
tweaks + lint

jonathanc-n commented:
@git-hulk @PragmaTwice I think a workflow can be run again with the changes

jonathanc-n committed:
max memory checks for initbyprob command

jonathanc-n committed:
Merge branch 'unstable' into cms

jonathanc-n committed:
Merge branch 'unstable' into cms

jonathanc-n committed:
Merge branch 'unstable' into cms

jonathanc-n commented:
@PragmaTwice Would you be able to review this pr?

mapleFU commented on src/types/redis_cms.h:
```diff
@@ -0,0 +1,50 @@
+  [[nodiscard]] rocksdb::Status GetMetadata(engine::Context &ctx, const Slice &ns_key,
+                                            CountMinSketchMetadata *metadata);
+
+  // TODO (jonathanc-n)
```
todo what?

jonathanc-n replied:
@mapleFU the mergeuserkeys function

mapleFU replied:
```suggestion
  // TODO (jonathanc-n)
  // [[nodiscard]] rocksdb::Status mergeUserKeys(engine::Context &ctx, const std::vector<Slice> &user_keys,
  //                                            std::vector<std::string> *register_segments);  
```

Can you do like this?

mapleFU commented on src/types/cms.h:
```diff
@@ -0,0 +1,85 @@
+#include <vector>
+
+#include "server/redis_reply.h"
+#include "xxhash.h"
```
I didn't see xxhash being used in header...

mapleFU commented on src/types/cms.h:
```diff
@@ -0,0 +1,85 @@
+
+class CMSketch {
+ public:
+  explicit CMSketch(uint32_t width, uint32_t depth, uint64_t counter, std::vector<uint32_t> array = {})
```
Please not use default argument, which might being misused when we program later

mapleFU commented on src/types/cms.cc:
```diff
@@ -0,0 +1,128 @@
+
+int CMSketch::Merge(CMSketch* dest, size_t quantity, const std::vector<const CMSketch*>& src,
+                    const std::vector<long long>& weights) {
+  if (checkOverflow(dest, quantity, src, weights) != 0) {
+    return -1;
+  }
```
Should we use `StatusOr<>` rather than `-1` code here?

jonathanc-n replied:
I would rather avoid using statusor on the implementation layer, dealing with future errors gets difficult to track.

mapleFU replied:
Why? So you think integers are better to tracking than Explicit Status code?

jonathanc-n replied:
Oh sorry, I wasn't aware of what Status was able to do.

mapleFU commented on src/types/cms.cc:
```diff
@@ -0,0 +1,128 @@
+    for (size_t j = 0; j < dest->GetWidth(); ++j) {
+      int64_t item_count = 0;
+      for (size_t k = 0; k < quantity; ++k) {
+        item_count += static_cast<int64_t>(src[k]->array_[(i * dest->GetWidth()) + j]) * weights[k];
```
`GetLocation * weights[k]`?

jonathanc-n replied:
No, i believe they dont do the same

mapleFU replied:
You're right, this avoid calculation the hash location here. I'm thinking that would `GetLocation -> GetLocationForHash` would being a less confusing naming

mapleFU commented on src/types/cms.cc:
```diff
@@ -0,0 +1,128 @@
+  return CMSketch::Merge(params.dest, params.num_keys, params.cms_array, params.weights);
+}
+
+int CMSketch::checkOverflow(CMSketch* dest, size_t quantity, const std::vector<const CMSketch*>& src,
```
Also using `StatusOr` here?

mapleFU reviewed (commented):
General LGTM
Would you mind also add the go test, for:
1. do operations on non-created
2. Basic query
3. Merge
?

jonathanc-n committed:
Update src/types/redis_cms.h

Co-authored-by: mwish <maplewish117@gmail.com>

jonathanc-n committed:
all fixes + go test case

jonathanc-n committed:
Merge branch 'unstable' into cms

jonathanc-n commented:
@mapleFU something seems to be a little odd here, I ran clang-format on all changed files and its running into formatting issues

jonathanc-n committed:
Merge branch 'unstable' into cms

jonathanc-n committed:
Merge branch 'unstable' into cms

mapleFU commented on tests/gocase/unit/cms/cms_test.go:
```diff
@@ -0,0 +1,245 @@
+	// Query for non-existent element
+	t.Run("cms.query - Query Non-Existent Element", func(t *testing.T) {
+		require.NoError(t, rdb.Do(ctx, "DEL", "cmsA").Err())
+		res := rdb.Do(ctx, "cms.initbydim", "cmsA", "100", "10")
+		require.NoError(t, res.Err())
+		require.Equal(t, "OK", res.Val())
```
Can we also query a non-exists key like "cms.incrby - Increment Non-Existent CMS"

mapleFU commented on tests/gocase/unit/cms/cms_test.go:
```diff
@@ -0,0 +1,245 @@
+		require.Equal(t, "0", count, "Non-existent element should return count '0'")
+	})
+
+	// Merging CMS structures
```
Can we also merge on unexists key? Includling uninitialized dest or sources

mapleFU reviewed (commented):
@jonathanc-n would you mind me add some comments and edit it directly after test is added?

jonathanc-n committed:
Added additional test cases

jonathanc-n committed:
lint fix

jonathanc-n commented:
@mapleFU Test cases are implemented! Ty for the test suggestions.

jonathanc-n commented:
@mapleFU Does this seem fine to move forward?

jonathanc-n committed:
Merge branch 'unstable' into cms

mapleFU commented on src/commands/cmd_cms.cc:
```diff
@@ -0,0 +1,266 @@
+
+    s = cms.IncrBy(ctx, args_[1], elements);
+    if (s.IsNotFound()) {
+      return {Status::RedisExecErr, "Key not found"};
```
Not Found handling is different from `CMS.INFO`?

mapleFU commented on src/commands/cmd_cms.cc:
```diff
@@ -0,0 +1,266 @@
+ private:
+  std::string destination_;  // Changed to std::string
+  int num_keys_;
+  std::vector<std::string> src_keys_;  // Changed to std::vector<std::string>
```
Actually I don't understanding why change to std::string...

mapleFU commented on src/types/redis_cms.h:
```diff
@@ -0,0 +1,48 @@
+  rocksdb::Status InitByDim(engine::Context &ctx, const Slice &user_key, uint32_t width, uint32_t depth);
+  rocksdb::Status InitByProb(engine::Context &ctx, const Slice &user_key, double error, double delta);
+  rocksdb::Status Query(engine::Context &ctx, const Slice &user_key, const std::vector<std::string> &elements,
+                        std::vector<uint32_t> &counters);
```
```suggestion
                        std::vector<uint32_t> *counters);
```

mapleFU commented on src/types/redis_cms.cc:
```diff
@@ -0,0 +1,264 @@
+  if (s.IsNotFound()) {
+    return rocksdb::Status::NotFound();
+  }
+  if (!s.ok() && !s.IsNotFound()) {
```
Why `!s.IsNotFound()` here?

mapleFU commented on src/types/redis_cms.cc:
```diff
@@ -0,0 +1,264 @@
+
+  CMSketch cms(metadata.width, metadata.depth, metadata.counter, metadata.array);
+
+  if (elements.empty()) {
+    return rocksdb::Status::OK();
+  }
```
I don't understand why it would be empty: https://github.com/RedisBloom/RedisBloom/blob/b5e674f61e84f9636b26d2005fe5ea50230ecb79/src/rm_cms.c#L120

mapleFU commented on src/types/redis_cms.cc:
```diff
@@ -0,0 +1,264 @@
+  }
+
+  for (auto &element : elements) {
+    cms.IncrBy(element.first, element.second);
```
We don't check "INCRBY overflow"?

mapleFU commented on src/types/redis_cms.cc:
```diff
@@ -0,0 +1,264 @@
+  CountMinSketchMetadata metadata{};
+  rocksdb::Status s = GetMetadata(ctx, ns_key, &metadata);
+
+  if (!s.ok() || s.IsNotFound()) {
```
I don't understand why `!s.ok()` can be cast to `NotFound`

mapleFU commented on src/types/redis_cms.cc:
```diff
@@ -0,0 +1,264 @@
+  rocksdb::Status s = GetMetadata(ctx, ns_key, &metadata);
+
+  if (!s.IsNotFound()) {
+    return s;
```
Shouldn't we return "key already exists" when `s.ok()`. If we follow the current way, it will return ok when init.

mapleFU commented on src/types/redis_cms.cc:
```diff
@@ -0,0 +1,264 @@
+};
+
+rocksdb::Status CMS::InitByProb(engine::Context &ctx, const Slice &user_key, double error, double delta) {
+  if (error <= 0 || error >= 1) {
+    return rocksdb::Status::InvalidArgument("Error must be between 0 and 1 (exclusive).");
+  }
+  if (delta <= 0 || delta >= 1) {
+    return rocksdb::Status::InvalidArgument("Delta must be between 0 and 1 (exclusive).");
+  }
```
Can we count a `(dim, ...)` then call `InitByDim`?

mapleFU commented on src/types/redis_cms.cc:
```diff
@@ -0,0 +1,264 @@
+}
+
+rocksdb::Status CMS::Query(engine::Context &ctx, const Slice &user_key, const std::vector<std::string> &elements,
+                           std::vector<uint32_t> &counters) {
```
Should we clear the counters firstly?

mapleFU reviewed (commented):
I found some unsolved issue marked as "solved", I spend some time on tracking them...

mapleFU commented on src/types/cms.h:
```diff
@@ -0,0 +1,84 @@
+    std::vector<long long> weights;
+  };
+
+  int CMSMergeParams(const MergeParams& params);
```
Ever used?

mapleFU commented on src/types/redis_cms.cc:
```diff
@@ -0,0 +1,264 @@
+
+  for (size_t i = 0; i < num_sources; ++i) {
+    std::string src_ns_key = AppendNamespacePrefix(src_keys[i]);
+    LOG(INFO) << "Dest Key: " << dest_ns_key << " | Source Key: " << src_ns_key;
```
Is this neccessary for INFO?

mapleFU commented on src/types/redis_cms.cc:
```diff
@@ -0,0 +1,264 @@
+  src_cms_objects.reserve(num_sources);
+  std::vector<const CMSketch *> src_cms_pointers;
+  src_cms_pointers.reserve(num_sources);
+  std::vector<long long> weights_long;
```
Should we uses `int64_t` for this?

jonathanc-n committed:
Update src/types/redis_cms.h

Co-authored-by: mwish <maplewish117@gmail.com>

jonathanc-n committed:
Fixes

jonathanc-n committed:
Merge branch 'cms' of https://github.com/jonathanc-n/kvrocks into cms

jonathanc-n commented:
@mapleFU Alright, thanks for the reviews, I put the changes in for it.

mapleFU commented on src/types/cms.cc:
```diff
@@ -0,0 +1,130 @@
+
+#include "cms.h"
+
+#include <xxhash.h>
```
wonder isn't xxhash include with `"`?

mapleFU reviewed (commented):
Rest LGTM. I'll update some comment before approve these two days

mapleFU commented on src/types/redis_cms.cc:
```diff
@@ -0,0 +1,251 @@
+  CountMinSketchMetadata metadata{};
+  rocksdb::Status s = GetMetadata(ctx, ns_key, &metadata);
+
+  if (!s.ok()) {
+    return rocksdb::Status::NotFound();
```
ditto, !ok doesn't means NotFound

jonathanc-n committed:
Small Changes

jonathanc-n committed:
logic fix

jonathanc-n committed:
lint

jonathanc-n commented:
@mapleFU Srorry about that, fixed the lint.

mapleFU commented:
No problem, we're almost their

mapleFU commented on src/types/redis_cms.cc:
```diff
@@ -0,0 +1,252 @@
+    return s;
+  }
+
+  size_t memory_used = dim.width * dim.depth * sizeof(uint32_t);
+  const size_t max_memory = 50 * 1024 * 1024;
```
Actually I think this should be put in `InitByDim` before allocating the memory or put to storage

PragmaTwice commented on src/types/cms.h:
```diff
@@ -0,0 +1,81 @@
+  uint64_t counter_;
+  std::vector<uint32_t> array_;
+
+  static int checkOverflow(CMSketch* dest, size_t quantity, const std::vector<const CMSketch*>& src,
+                           const std::vector<long long>& weights);
```
Why don't we use `this` and make it a member function here? instead of `dest`.

PragmaTwice commented on src/types/cms.h:
```diff
@@ -0,0 +1,81 @@
+    std::vector<long long> weights;
+  };
+
+  static Status Merge(const MergeParams& params);
```
Maybe we can just use these fields of MergeParams as parameters? i.e. `Merge(size_t num_keys, ...)`.

jonathanc-n committed:
Quick Fixes

mapleFU commented:
I'm a bit tired today and will revisit this tomorrow

jonathanc-n committed:
lint fix

jonathanc-n committed:
typing fixes

PragmaTwice commented on src/types/cms.h:
```diff
@@ -0,0 +1,82 @@
+
+  size_t Query(std::string_view item) const;
+
+  struct MergeParams {
+    CMSketch* dest;
+    size_t num_keys;
+    std::vector<const CMSketch*> cms_array;
+    std::vector<uint32_t> weights;
+  };
```
seems useless now

jonathanc-n committed:
one mb limit update per key

PragmaTwice commented on src/types/redis_cms.cc:
```diff
@@ -0,0 +1,256 @@
+    return rocksdb::Status::InvalidArgument("Memory usage must be greater than 0.");
+  }
+  if (memory_used > max_memory) {
+    return rocksdb::Status::InvalidArgument("Memory usage exceeds 50MB.");
```
```suggestion
    return rocksdb::Status::InvalidArgument("Memory usage exceeds 1MB.");
```

PragmaTwice commented on src/types/redis_cms.cc:
```diff
@@ -0,0 +1,256 @@
+    return rocksdb::Status::InvalidArgument("Memory usage must be greater than 0.");
+  }
+  if (memory_used > max_memory) {
+    return rocksdb::Status::InvalidArgument("Memory usage exceeds 50MB.");
```
```suggestion
    return rocksdb::Status::InvalidArgument("Memory usage exceeds 1MB.");
```

jonathanc-n committed:
Update src/types/redis_cms.cc

Co-authored-by: Twice <twice@apache.org>

jonathanc-n committed:
Update src/types/redis_cms.cc

Co-authored-by: Twice <twice@apache.org>

mapleFU committed:
[WIP] Some codereview check

1. Add just IncrBy syntax (wip)
2. Extract a hash function rather than explicit xxh
3. Fix a bug in merge

mapleFU commented:
I just reach home and have some minor updates, would continue tomorrow. Sorry for delaying

mapleFU commented on src/types/redis_cms.cc:
```diff
@@ -204,11 +191,15 @@ rocksdb::Status CMS::MergeUserKeys(engine::Context &ctx, const Slice &user_key,
     weights_long.push_back(static_cast<uint32_t>(src_weights[i]));
   }
-
+  // Initialize the destination CMS with the source CMSes after initializations
+  // since vector might resize and reallocate memory.
```
Previously this might cause memory issue

mapleFU commented on src/types/redis_cms.cc:
```diff
@@ -165,33 +151,34 @@ rocksdb::Status CMS::MergeUserKeys(engine::Context &ctx, const Slice &user_key,
   }
 
   std::string dest_ns_key = AppendNamespacePrefix(user_key);
-  LockGuard guard(storage_->GetLockManager(), dest_ns_key);
```
This might cause deadlock

jonathanc-n commented:
@mapleFU Were you gonna make your own changes to the code?

aleksraiden committed:
Merge branch 'unstable' into cms

mapleFU commented:
Yes I'm back from holiday and working, would continue.

mapleFU committed:
Update: making content pass test
